fix(core): Store ICU state in `LView` rather than in `TView`

mhevery opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

mhevery committed:
fix(core): Store ICU state in `LView` rather than in `TView`

Before this refactoring/fix the ICU would store the current selected
index in `TView`. This is incorrect, since if ICU is in `ngFor` it will
cause issues in some circumstances. This refactoring properly moves the
state to `LView`.

closes #37021
closes #38144
closes #38073

mhevery committed:
refactor(core): Change `TemplateFixture` to named parameters

`TemplateFixture` used to have positional parameters and many tests got
hard to read as number of parameters reach 10+ with many of them `null`.
This refactoring changes `TemplateFixture` to take named parameters
which improves usability and readability in tests.

alfaproject commented on packages/core/src/render3/instructions/lview_debug.ts:
```diff
@@ -170,6 +170,7 @@ class TNode implements ITNode {
       public tView_: TView,                                                          //
       public type: TNodeType,                                                        //
       public index: number,                                                          //
+      public insertBeforeIndex: null|- 1 | number | number[],                        //
```
`-1 | number` is `number`

mhevery replied:
Very true! But I like to add `-1` to let the world know that it has special meaning.
In either case the code has been refactored and `-1` is no longer needed at this location.

mhevery committed:
refactor(core): move `i18n_spec.ts` into `i18n` subfolder

`i18n_spec.ts` file was incorrectly in the `render3` folder rather than `render3/i18n`

mhevery commented:
[presubmit](http://test/OCL:336176711:BASE:336176744:1602195937294:46e2d068)

mhevery commented:
[presubmit](http://test/OCL:336385739:BASE:336385758:1602285680732:2e434b63)

petebacondarwin commented on packages/core/test/render3/render_util.ts:
```diff
@@ -105,12 +109,23 @@ export class TemplateFixture extends BaseFixture {
-      rendererFactory?: RendererFactory3, private _consts?: TConstants) {
+  constructor({create, update, decls, vars, directives, pipes, sanitizer, rendererFactory, consts}:
+                  {
+                    create?: (() => void)|null,
```
If there a difference between `undefined` (i.e. the `create` is not provided in the call) and `null`, which is part of the type? Further down there is:

```ts
this.createBlock = create || noop;
```

So if `create` is either `undefined` or `null` you will end up with `this.createBlock` being `noop`.

This is different to the previous behaviour of this constructor.

(Similarly for `update`.

mhevery replied:
no difference. cleaned up.

petebacondarwin commented on packages/core/test/render3/render_util.ts:
```diff
@@ -105,12 +109,23 @@ export class TemplateFixture extends BaseFixture {
+                  }) {
     super();
+    this._consts = consts || [];
+    this._vars = vars || 0;
```
Obviously this is not a problem now, but if you decided to change the default value to say `2` then a value of 0 for `consts` would get changed to `2`...

One option is to use `??` instead of `||`.

Another would be to move all these defaults into the call signature:

```ts
constructor({create = noop, update = noop, decls, vars = 0, directives, pipes, sanitizer = null, rendererFactory, consts = []}:
```

mhevery replied:
I did not know that you can defaults with parameterized types. Fixed it.

petebacondarwin commented on packages/core/test/render3/render_util.ts:
```diff
@@ -125,8 +140,8 @@ export class TemplateFixture extends BaseFixture {
         },
-        decls, vars, null!, this._rendererFactory, null, this._directiveDefs, this._pipeDefs,
-        sanitizer, this._consts);
+        decls || 0, vars || 0, null!, this._rendererFactory, null, this._directiveDefs,
```
Moving the default initialization into the signature would avoid the need for these `||` fallbacks here.

mhevery replied:
done

petebacondarwin reviewed (commented):
First commit:

petebacondarwin reviewed (commented):
2nd commit: why does this spec need to be in a subfolder, when lots of other specs are still in the `render3` folder? Also should `packages/core/test/render3/i18n_debug_spec.ts` move to the `i18n` as well in that case?

mhevery commented on packages/core/src/util/assert.ts:
```diff
@@ -111,6 +111,9 @@ export function assertDomNode(node: any): asserts node is Node {
-  assertLessThan(index, maxLen, `Index expected to be less than ${maxLen} but got ${index}`);
+  assertDefined(arr, 'Array must be defined.');
+  const maxLen = arr.length;
+  if (index < 0 || index > maxLen) {
```
yes, it should.! fixed.

AndrewKushnir commented on packages/core/src/render3/component.ts:
```diff
@@ -174,7 +174,7 @@ export function createRootComponentView(
   ngDevMode && assertIndexInRange(rootView, 0 + HEADER_OFFSET);
   rootView[0 + HEADER_OFFSET] = rNode;
-  const tNode: TElementNode = getOrCreateTNode(tView, 0, TNodeType.Element, null, null);
+  const tNode: TElementNode = getOrCreateTNode(tView, 0, TNodeType.Element, '#host', null);
```
It looks this change is not needed (and should be reverted)?

AndrewKushnir reviewed (commented):
@mhevery I've posted first few comments, will have more feedback tomorrow.

petebacondarwin commented:
Sorry to be a pain, @mhevery , but the changes you made to resolve issues in the first commit got amended to the last commit, which is problematic.

The original commit is still broken, so if the last commit got reverted the first commit would be wrong.

Also, it is hard to see how the review comments were resolved. This is especially significant in this PR since the last commit is already incredibly large. It would have been much nicer to land those resolutions to the comments in a fixup commit that referenced the first commit where the issue was.

alfaproject commented on packages/core/src/render3/i18n/i18n_apply.ts:
```diff
@@ -506,5 +425,5 @@ function getCaseIndex(icuExpression: TIcu, bindingValue: string): number {
     }
   }
-  return index;
+  return index === -1 ? null : index;
```
It doesn't look like you need to change the return type, right? (also returning `-1` is fairly standard for index retrieving methods)

petebacondarwin commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -606,23 +633,67 @@ function getNativeAnchorNode(parentTNode: TNode, lView: LView): RNode|null {
  *
```
The line above this mentions `canInsertNativeNode` and insertion delay, both of which I believe no longer exist, right?

petebacondarwin commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -606,23 +633,67 @@ function getNativeAnchorNode(parentTNode: TNode, lView: LView): RNode|null {
+    }
+  }
+
+  const tNodeInsertBeforeIndex = childTNode.insertBeforeIndex;
+  if (Array.isArray(tNodeInsertBeforeIndex)) {
+    // Array means that we have a special place we need to insert in front of as well as append
+    // children to ourselves. See `TNode.insertBeforeIndex` for context.
+    processI18nText(renderer, childTNode, lView, childRNode, renderParent, tNodeInsertBeforeIndex);
+  }
```
Is the idea here that we are "now" adding the nodes that were previously placeholders? And so now is the time to wire up the child i18n nodes that were waiting for their real DOM node to be provided?

petebacondarwin replied:
How about:

```
// An array indicates that there are i18n nodes that need to be added as children of this `rChildNode` node.
// These i18n nodes were created before this `rChildNode` was available and so only now can be added.
// The first element of the array is the normal index where we should insert the `rChildNode`.
/// Additional elements are the extra nodes to be added as children of `rChildNode`.
```

mhevery replied:
done

petebacondarwin commented on packages/core/test/acceptance/i18n_spec.ts:
```diff
@@ -633,183 +631,58 @@ onlyInIvy('Ivy i18n logic').describe('runtime i18n', () => {
-      });
-
-      // FIXME(misko): re-enable and fix this use case.
-      xit('should support multiple ICUs', () => {
```
🎉

petebacondarwin commented on packages/core/test/acceptance/i18n_spec.ts:
```diff
@@ -1061,27 +946,27 @@ onlyInIvy('Ivy i18n logic').describe('runtime i18n', () => {
-      //   if (i === 1 && v === 0) return 1;
-      //   if (!(v === 0) || n === 0 ||
-      //       !(n === 1) && n % 100 === Math.floor(n % 100) && n % 100 >= 1 && n % 100 <= 19)
+      //   let i = Math.floor(Math.abs(n)), v = n.toString().replace(/^[^.]*\.?/,
+      //   '').length; if (i === 1 && v === 0) return 1; if (!(v === 0) || n === 0 ||
+      //       !(n === 1) && n % 100 === Math.floor(n % 100) && n % 100 >= 1 && n % 100 <=
+      //       19)
```
Ideally we should turn off code formatting on this comment to prevent it from treating it like prose.
Perhaps wrapping it in triple backticks might be enough?

petebacondarwin replied:
Actually, the current code formatter (`yarn ng-dev format changed`) doesn't wrap these code lines.
So perhaps just revert this change?

mhevery replied:
reverted

petebacondarwin commented on packages/core/test/acceptance/i18n_spec.ts:
```diff
@@ -14,9 +14,7 @@ import localeEs from '@angular/common/locales/es';
-import {getComponentDef} from '@angular/core/src/render3/definition';
-import {setDelayProjection} from '@angular/core/src/render3/instructions/projection';
-import {TI18n, TIcu} from '@angular/core/src/render3/interfaces/i18n';
+import {TI18n} from '@angular/core/src/render3/interfaces/i18n';
```
This import is not used.

mhevery replied:
👍

petebacondarwin commented on packages/core/test/acceptance/i18n_spec.ts:
```diff
@@ -14,9 +14,7 @@ import localeEs from '@angular/common/locales/es';
-import {setDelayProjection} from '@angular/core/src/render3/instructions/projection';
-import {TI18n, TIcu} from '@angular/core/src/render3/interfaces/i18n';
+import {TI18n} from '@angular/core/src/render3/interfaces/i18n';
 import {DebugNode, HEADER_OFFSET, TVIEW} from '@angular/core/src/render3/interfaces/view';
```
TVIEW is not used

mhevery replied:
👍

petebacondarwin commented on packages/core/test/acceptance/i18n_spec.ts:
```diff
@@ -14,9 +14,7 @@ import localeEs from '@angular/common/locales/es';
-import {TI18n, TIcu} from '@angular/core/src/render3/interfaces/i18n';
+import {TI18n} from '@angular/core/src/render3/interfaces/i18n';
 import {DebugNode, HEADER_OFFSET, TVIEW} from '@angular/core/src/render3/interfaces/view';
 import {getComponentLView, loadLContext} from '@angular/core/src/render3/util/discovery_utils';
```
`loadContext` is not used

mhevery replied:
👍

petebacondarwin commented on packages/core/test/acceptance/i18n_spec.ts:
```diff
@@ -2850,6 +2728,39 @@ onlyInIvy('Ivy i18n logic').describe('runtime i18n', () => {
+        // This test demonstrates an issue with creating a `ViewContainerRef` and having i18n at the
+        // parent element. The reason this broke is that in this case the `ViewContainerRef` creates
+        // an dynamic anchor comment but uses `HostTNode` for it which is incorrect. `appendChild`
+        // then tries to add internationalization to the comment node and fails.
```
Perhaps add a link to the issue(s) that reported this?

mhevery replied:
Hmm, no easy way to do that. It was just presubmit and those are only kept for 30 days.

petebacondarwin commented on packages/core/test/acceptance/i18n_spec.ts:
```diff
@@ -2850,6 +2728,39 @@ onlyInIvy('Ivy i18n logic').describe('runtime i18n', () => {
     });
   });
+
+  describe('regression', () => {
```
Is the `regression` describe block necessary?

mhevery replied:
removed

petebacondarwin commented on packages/core/test/render3/i18n/i18n_insert_before_index_spec.ts:
```diff
@@ -0,0 +1,183 @@
+  });
+
+  describe('when adding a placeholder', () => {
+    describe('which is large index', () => {
```
```suggestion
    describe('whose index is greater than those already there', () => {
```

mhevery replied:
👍

petebacondarwin commented on packages/core/test/render3/i18n/i18n_insert_before_index_spec.ts:
```diff
@@ -0,0 +1,183 @@
+
+  describe('when adding a placeholder', () => {
+    describe('which is large index', () => {
+      it('should do nothing', () => {
```
```suggestion
      it('should not update the `insertBeforeIndex` values', () => {
```

mhevery replied:
👍

petebacondarwin commented on packages/core/test/render3/i18n/i18n_insert_before_index_spec.ts:
```diff
@@ -0,0 +1,183 @@
+      });
+    });
+
+    describe('which is smaller index', () => {
```
```suggestion
    describe('whose index is smaller than current nodes', () => {
```

mhevery replied:
👍

petebacondarwin commented on packages/core/test/render3/i18n/i18n_insert_before_index_spec.ts:
```diff
@@ -0,0 +1,183 @@
+
+
+describe('addTNodeAndUpdateInsertBeforeIndex', () => {
+  function tNode(index: number, type: TNodeType, insertBeforeIndex: number|null = null): TNode {
+    const tNode = createTNode(null!, null, type, index, null, null);
+    tNode.insertBeforeIndex = insertBeforeIndex;
+    return tNode;
+  }
+
+  function tPlaceholderElementNode(index: number, insertBeforeIndex: number|null = null) {
+    return tNode(index, TNodeType.Placeholder, insertBeforeIndex);
+  }
+
+  function tI18NTextNode(index: number, insertBeforeIndex: number|null = null) {
+    return tNode(index, TNodeType.Element, insertBeforeIndex);
+  }
```
Really nice helpers!

mhevery replied:
❤️

petebacondarwin commented on packages/core/test/render3/i18n/i18n_insert_before_index_spec.ts:
```diff
@@ -0,0 +1,183 @@
+  });
+
+  describe('when adding a i18nText', () => {
+    describe('which is large index', () => {
```
```suggestion
    describe('whose index is greater than those already there', () => {
```

mhevery replied:
👍

petebacondarwin commented on packages/core/test/render3/i18n/i18n_insert_before_index_spec.ts:
```diff
@@ -0,0 +1,183 @@
+
+  describe('when adding a i18nText', () => {
+    describe('which is large index', () => {
+      it('should do nothing', () => {
```
```suggestion
      it('should not update the `insertBeforeIndex` values', () => {
```

petebacondarwin commented on packages/core/test/render3/i18n/i18n_insert_before_index_spec.ts:
```diff
@@ -0,0 +1,183 @@
+      });
+    });
+
+    describe('which is smaller index', () => {
```
```suggestion
    describe('whose index is smaller than current nodes', () => {
```

mhevery replied:
👍

petebacondarwin commented on packages/core/test/render3/i18n/i18n_spec.ts:
```diff
@@ -60,39 +63,41 @@ describe('Runtime i18n', () => {
     return new TemplateFixture({
       create: createTemplate,
-      update: updateTemplate,
+      update: updateTemplate || undefined,
```
Instead of this you can make the parameter `updateTemplate: (() => void)|undefined` and then proliferate that type change through the rest of the file.

mhevery replied:
👍

petebacondarwin commented on packages/core/test/render3/i18n/i18n_spec.ts:
```diff
@@ -60,39 +63,41 @@ describe('Runtime i18n', () => {
         create: debugMatch([
-          'lView[1] = document.createTextNode("simple text")',
-          '(lView[0] as Element).appendChild(lView[1])'
+          'lView[21] = document.createText("simple text");',
```
How is `21` calculated? I believe it is `HEADER_OFFSET + 1` right? If so it might be worth doing this computation here both to make it more clear to the reader but also to make it more resilient to changes to `HEADER_OFFSET`.

Similarly elsewhere in this file.

mhevery replied:
done

petebacondarwin commented on packages/core/test/render3/i18n/i18n_spec.ts:
```diff
@@ -641,4 +636,127 @@ describe('Runtime i18n', () => {
+        i18nRangeOffsetOpcode(0), '',  //
+        // 1 is saved for binding
+        i18nRangeOffsetOpcode(1, {appendLater: true}), '',  //
+        // 3 is saved for binding
```
Is this comment correct?

mhevery replied:
yes it is. Reformated to make it more clear.

petebacondarwin reviewed (commented):
Amazing PR @mhevery. I took me about 8 hours to look at it all.
I left lots of comments but I guess that is because there are a lot of code changes!

AndrewKushnir commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -588,13 +605,23 @@ export function nativeNextSibling(renderer: Renderer3, node: RNode): RNode|null
-      parentTNode.type === TNodeType.IcuContainer) {
-    return getNativeByTNode(parentTNode, lView);
+function getNativeAnchorNode(parentTNode: TNode, currentTNode: TNode, lView: LView): RNode|null {
+  const tNodeInsertBeforeIndex = currentTNode.insertBeforeIndex;
+  const insertBeforeIndex =
+      Array.isArray(tNodeInsertBeforeIndex) ? tNodeInsertBeforeIndex[0] : tNodeInsertBeforeIndex;
```
We can probably create a helper method for this logic as it seems to be used in multiple places:

```typescript
function getInsertBeforeIndex(tNode: TNode): number {
  const index = currentTNode.insertBeforeIndex;
  return Array.isArray(index) ? index[0] : index;
}

const insertBeforeIndex = getInsertBeforeIndex(currentTNode);
```

AndrewKushnir commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -606,23 +633,67 @@ function getNativeAnchorNode(parentTNode: TNode, lView: LView): RNode|null {
+ */
+function processI18nText(
+    renderer: Renderer3, childTNode: TNode, lView: LView, childRNode: RNode|RNode[],
+    renderParent: RElement|null, i18nChildren: number[]): void {
```
```suggestion
    parentRNode: RElement|null, i18nChildren: number[]): void {
```

AndrewKushnir replied:
It looks like the `renderParent` was not renamed...

mhevery replied:
sorry, not sure how I missed it. Fixed now.

AndrewKushnir commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -606,23 +633,67 @@ function getNativeAnchorNode(parentTNode: TNode, lView: LView): RNode|null {
   const renderParent = getRenderParent(tView, childTNode, lView);
+  const renderer = lView[RENDERER];
+  const parentTNode: TNode = childTNode.parent || lView[T_HOST]!;
+  const anchorNode = getNativeAnchorNode(parentTNode, childTNode, lView);
```
```suggestion
  const anchorRNode = getNativeAnchorNode(parentTNode, childTNode, lView);
```

AndrewKushnir commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -606,23 +633,67 @@ function getNativeAnchorNode(parentTNode: TNode, lView: LView): RNode|null {
 export function appendChild(
-    tView: TView, lView: LView, childEl: RNode|RNode[], childTNode: TNode): void {
+    tView: TView, lView: LView, childRNode: RNode|RNode[], childTNode: TNode): void {
   const renderParent = getRenderParent(tView, childTNode, lView);
```
```suggestion
  const parentRNode = getParentRNode(tView, childTNode, lView);
```

AndrewKushnir commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -606,23 +633,67 @@ function getNativeAnchorNode(parentTNode: TNode, lView: LView): RNode|null {
+      const i18nChild = lView[i18nChildren[i]];
+      ngDevMode && assertDomNode(i18nChild);
+      isProcedural ?
+          (renderer as ProceduralRenderer3).insertBefore(i18nParent, i18nChild, anchorNode) :
```
We can use `nativeInsertBefore` function (located in this file) here.

AndrewKushnir commented on packages/core/src/render3/i18n/i18n_parse.ts:
```diff
@@ -33,22 +36,9 @@ const ICU_REGEXP = /({\s*�\d+:?\d*�\s*,\s*\S{6}\s*,[\s\S]*})/gi;
-
 const MARKER = `�`;
 const SUBTEMPLATE_REGEXP = /�\/?\*(\d+:\d+)�/gi;
 const PH_REGEXP = /�(\/?[#*!]\d+):?\d*�/gi;
```
Right, I have a small PR that handles that: #39172, but we can just incorporate the changes into this PR.

AndrewKushnir commented on packages/core/src/render3/i18n/i18n_parse.ts:
```diff
@@ -62,148 +52,186 @@ function replaceNgsp(value: string): string {
+  const i18nNodeIdx = allocExpando(tView, lView, 1);
+  let opCode = i18nNodeIdx << I18nCreateOpCode.SHIFT;
+  const previousOrParent = getCurrentTNode()!;
+  const isParent = isCurrentTNodeParent();
+  let parentTNode = isParent ? previousOrParent : previousOrParent.parent;
```
This can probably be replaced by `getCurrentParentTNode` call:
```suggestion
  let parentTNode = getCurrentParentTNode();
```

AndrewKushnir commented on packages/core/src/render3/i18n/i18n_parse.ts:
```diff
@@ -266,15 +295,22 @@ export function i18nAttributesFirstPass(
-    str: string, destinationNode: number, attrName?: string,
-    sanitizeFn: SanitizerFn|null = null): I18nUpdateOpCodes {
-  const updateOpCodes: I18nUpdateOpCodes = [null, null];  // Alloc space for mask and size
+    updateOpCodes: I18nUpdateOpCodes, str: string, destinationNode: number, attrName?: string,
```
The `attrName` can not be optional, since the next argument is not optional too, the proposed update is:
```suggestion
    updateOpCodes: I18nUpdateOpCodes, str: string, destinationNode: number, attrName: string|undefined,
```

AndrewKushnir commented on packages/core/src/render3/i18n/i18n_parse.ts:
```diff
@@ -413,19 +444,37 @@ export function getTranslationForTemplate(message: string, subTemplateIndex?: nu
+  const tIcu: TIcu = {
+    type: icuExpression.type,
+    currentCaseLViewIndex: allocExpando(tView, lView, 1),
+    anchorIdx: anchorIdx,
```
```suggestion
    anchorIdx,
```

mhevery commented on packages/core/test/render3/render_util.ts:
```diff
@@ -105,16 +110,36 @@ export class TemplateFixture extends BaseFixture {
+    this.updateBlock = update;
     this._directiveDefs = toDefs(directives, extractDirectiveDef);
     this._pipeDefs = toDefs(pipes, extractPipeDef);
     this._sanitizer = sanitizer || null;
```
fixed

mhevery commented on packages/core/src/render3/instructions/advance.ts:
```diff
@@ -6,6 +6,7 @@
  * found in the LICENSE file at https://angular.io/license
  */
 import {assertGreaterThan, assertIndexInRange} from '../../util/assert';
```
fixed

mhevery commented on packages/core/test/render3/view_utils_spec.ts:
```diff
@@ -6,22 +6,21 @@
+import {ViewFixture} from './view_fixture';
 
 describe('view_utils', () => {
   it('should verify unwrap methods', () => {
```
fixed

AndrewKushnir commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -230,14 +215,27 @@ export function getOrCreateTNode(
+    tNode = createTNodeAtIndex(tView, adjustedIndex, type, name, attrs);
+    if (isInI18nBlock()) {
+      // If we are in i18n block then all elements should be pre declared through `Placeholder`
+      // See `TNodeType.Placeholder` and `LFrame.inI18n` for more context.
```
nit: it'd be great to mention why we create detached `tNode`s at all.

mhevery replied:
clarified

AndrewKushnir commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -1506,15 +1519,20 @@ export function elementAttributeInternal(
+  setElementAttribute(lView[RENDERER], element, namespace, tNode.tagName, name, value, sanitizer);
+}
+
+export function setElementAttribute(
```
It looks like we can just move the logic back to the `elementAttributeInternal` function and use it in `i18n_apply.ts` directly (we use it there already). Currently the `setElementAttribute` don't have asserts that the `elementAttributeInternal` has.

mhevery replied:
Unfortunately we can't. The `elementAttributeInternal` wants `TNode` which does not exists for the ICU case.

AndrewKushnir commented on packages/core/src/util/assert.ts:
```diff
@@ -102,15 +102,27 @@ export function throwError(msg: string, actual?: any, expected?: any, comparison
+
+
+export function assertOneOf(value: any, ...validValues: any[]) {
+  for (let index = 0; index < validValues.length; index++) {
+    if (validValues.indexOf(value) !== -1) return true;
+  }
```
Can it just be this (it looks like we don't need a `for` loop):
```suggestion
  return validValues.indexOf(value) !== -1;
```

mhevery replied:
opps, sorry about that.

AndrewKushnir commented on packages/core/src/render3/i18n/i18n_apply.ts:
```diff
@@ -7,132 +7,191 @@
-}
 
+/**
+ * Keep track of which input biddings in `ɵɵi18nExp` have changed.
```
```suggestion
 * Keep track of which input bindings in `ɵɵi18nExp` have changed.
```

mhevery replied:
fixed

AndrewKushnir commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -93,13 +94,37 @@ function applyToElementOrContainer(
 
+export function createCommentNode(renderer: Renderer3, value: string): RComment {
+  ngDevMode && ngDevMode.rendererCreateComment++;
+  return renderer.createComment(value);
```
I'd propose to add a comment here to mention that the Render2 and Render3 APIs are the same when it comes to comment node creation. Since we have the `isProceduralRenderer(renderer)` check and 2 branches in most cases, it feel like it's missing here as well, when in fact it's not :)

mhevery replied:
added.

AndrewKushnir commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -656,6 +751,10 @@ function getFirstNativeNode(lView: LView, tNode: TNode|null): RNode|null {
+    } else if (tNodeType === TNodeType.IcuContainer) {
+      let nextRNode = icuContainerIterate(tNode as TIcuContainerNode, lView);
+      let rNode: RNode|null = nextRNode();
+      return rNode || unwrapRNode(lView[tNode.index]);
```
It'd be great to add a comment here to explain what happens when ICU container has no nodes.

mhevery replied:
good idea, done.

AndrewKushnir reviewed (approved):
LGTM, thanks @mhevery 👍

I've just left a few comments, but all of them are minor.

petebacondarwin commented on packages/core/test/render3/matchers.ts:
```diff
@@ -37,22 +37,20 @@ export function matchObjectShape<T>(
+      if (expected.hasOwnProperty(key) && !jasmine.matchersUtil.equals(_actual[key], expected[key]))
+        errors.push(
+            '\n  ' +
+            jasmine.matchersUtil.buildFailureMessage(
+                `obj.${key}`, false, _actual[key], expected[key]));
```
So regarding this comment: https://github.com/angular/angular/pull/39003/files?file-filters%5B%5D=.json&file-filters%5B%5D=.md&file-filters%5B%5D=.ts&hide-deleted-files=true#r504014425

There is no need to actually use `buildFailureMessage()` here. We can just generate our own message string. This helper is only useful to handle generic cases.

Something like:
```suggestion
            `property obj.${key} to equal ${expected[key]} but got ${_actual[key]}`);
```

mhevery replied:
fixed

petebacondarwin commented on packages/core/test/render3/matchers_spec.ts:
```diff
@@ -41,13 +41,8 @@ describe('render3 matchers', () => {
-          propA: "different"
-        }))`);
+      expect(matcher.jasmineToString!())
+          .toEqual('\n  Expected \'value\' obj.prop a \'different\'.');
```
See https://github.com/angular/angular/pull/39003/files?file-filters%5B%5D=.json&file-filters%5B%5D=.md&file-filters%5B%5D=.ts&hide-deleted-files=true#r505799571

petebacondarwin reviewed (approved):
LGTM code-wise.

mhevery commented on packages/core/src/render3/component.ts:
```diff
@@ -174,7 +175,7 @@ export function createRootComponentView(
   ngDevMode && assertIndexInRange(rootView, 0 + HEADER_OFFSET);
   rootView[0 + HEADER_OFFSET] = rNode;
-  const tNode: TElementNode = getOrCreateTNode(tView, 0, TNodeType.Element, null, null);
+  const tNode = getOrCreateTNode(tView, 0, TNodeType.Element, '#host', null);
```
added comment.

mhevery commented:
[presubmit](http://test/OCL:336385739:BASE:337538779:1602871206162:941ecfcd)

mhevery commented:
[presubmit](http://test/OCL:336385739:BASE:337568264:1602880488283:144678c2)

mhevery commented:
[presubmit](http://test/OCL:337604354:BASE:337604355:1602893314813:891907fe)

mhevery commented:
[presubmit](http://test/OCL:337604354:BASE:337610851:1602896845073:c69522fd)

mhevery commented:
[presubmit](http://test/OCL:337628320:BASE:337628331:1602911368624:b1dcb23c)

mhevery commented:
[presubmit](http://test/OCL:337628320:BASE:337674527:1602958903335:58f6c181)

mhevery commented:
[presubmit](http://test/OCL:337628320:BASE:337675941:1602961142970:84dfd1de)

mhevery commented:
CARETAKER: The failing test seems to be cause by https://github.com/angular/angular/pull/39308 and needs to be reverted. /cc @kyliau

petebacondarwin reviewed (approved):
Reviewed-for: global-docs-approvers

mhevery commented:
Closing in favor of https://github.com/angular/angular/pull/39233
